srml/im-online/src/lib.rs

```diff
@@ -0,0 +1,442 @@
+			}
+		}
+
+		// Runs after every block.
+		fn offchain_worker(now: T::BlockNumber) {
+			fn gossip_at<T: Trait>(block_number: T::BlockNumber) -> Result<(), OffchainErr> {
+				let kind = match <T::AuthorityId as TypedKey>::KEY_TYPE {
+					key_types::SR25519 => CryptoKind::Sr25519,
+					key_types::ED25519 => CryptoKind::Ed25519,
+					_ => return Err(OffchainErr::UnknownCryptoKind),
+				};
+
+				match sr_io::authority_pubkey(kind) {
+					Ok(key) => {
+						let authority_id = <T as Trait>::AuthorityId::decode(&mut &key[..])
+							.ok_or(OffchainErr::DecodeAuthorityId)?;
+						let network_state =
+							sr_io::network_state().map_err(|_| OffchainErr::NetworkState)?;
+						let heartbeat_data = Heartbeat {
+							block_number,
+							network_state,
+							session_index: <session::Module<T>>::current_index(),
+							authority_id,
+						};
+
+						match sr_io::sign(None, kind, &heartbeat_data.encode()) {
+							Ok(signature) => {
+								let call = Call::heartbeat(heartbeat_data, signature);
+								let ex = T::UncheckedExtrinsic::new_unsigned(call.into())
+									.ok_or(OffchainErr::ExtrinsicCreation)?;
+								sr_io::submit_transaction(&ex)
+									.map_err(|_| OffchainErr::SubmitTransaction)?;
+								set_worker_status::<T>(block_number, true);
+								Ok(())
+							},
+							Err(_) => Err(OffchainErr::FailedSigning),
+						}
+					},
+					Err(_) => {
+						// we only run when a local authority key is configured
+						Ok(())
+					}
+				}
+			}
+
+			fn set_worker_status<T: Trait>(gossipping_at: T::BlockNumber, done: bool) {
+				let enc = WorkerStatus {
+					done,
+					gossipping_at,
+				};
+				sr_io::local_storage_set(StorageKind::PERSISTENT, DB_KEY, &enc.encode());
+			}
+
+			fn was_not_yet_gossipped<T: Trait>(
+				now: T::BlockNumber,
+				next_gossip: T::BlockNumber,
+			) -> Result<bool, OffchainErr> {
+				let last_gossip = sr_io::local_storage_get(StorageKind::PERSISTENT, DB_KEY);
+				match last_gossip {
+					Some(l) => {
+						let worker_status: WorkerStatus<T::BlockNumber> = Decode::decode(&mut &l[..])
+							.ok_or(OffchainErr::DecodeWorkerStatus)?;
+
+						let was_aborted =
+							worker_status.done == false && worker_status.gossipping_at < now;
+
+						// another off-chain worker is currently in the process of submitting
+						let already_submitting =
+							worker_status.done == false && worker_status.gossipping_at == now;
+
+						let not_yet_gossipped =
+							worker_status.done == true && worker_status.gossipping_at < next_gossip;
+
+						let ret = (was_aborted == true && already_submitting == false) || not_yet_gossipped;
+						Ok(ret)
+					},
+					None => Ok(true),
+				}
+			}
+
+			let next_gossip = <GossipAt<T>>::get();
+			let not_yet_gossipped = match was_not_yet_gossipped::<T>(now, next_gossip) {
+				Ok(v) => v,
+				Err(err) => {
+					print(err);
+					return;
+				},
+			};
+			if next_gossip < now && not_yet_gossipped {
+				set_worker_status::<T>(now, false);
```

Comment:
This is subject to race conditions. I think we should use `compare_and_set` from offchain workers API.

Reply:
Created a follow-up (https://github.com/paritytech/substrate/issues/3143) and assigned to myself after private conversation.